src/Servers/Kestrel/Core/src/Internal/Http/HttpParser.cs

```diff
@@ -172,190 +185,132 @@ public bool ParseHeaders(TRequestHandler handler, ref SequenceReader<byte> reade
                     return false;
                 }
+                else if (span.Length == 0)
+                {
+                    KestrelBadHttpRequestException.Throw(RequestRejectionReason.InvalidRequestHeadersNoCRLF);
+                }
+                else
+                {
+                    RejectRequestHeader(span);
+                }
+            }
+            // No CR or LF
+            else
+            {
+                return false;
+            }
 
-                reader.Advance(length);
-                // As we crossed spans set the current span to default
-                // so we move to the next span on the next iteration
-                span = default;
+            if (!TryTakeSingleHeader(handler, span))
+            {
+                // Sequence needs to be CRLF and not contain an inner CR not part of terminator.
+                // Not parsable as a valid name:value header pair.
+                RejectRequestHeader(AppendEndOfLine(span, lineFeedOnly: eolSize == 1));
             }
         }
 
         return false;
     }
 
-    private int ParseMultiSpanHeader(TRequestHandler handler, ref SequenceReader<byte> reader)
+    private static byte[] AppendEndOfLine(ReadOnlySpan<byte> span, bool lineFeedOnly)
     {
-        var currentSlice = reader.UnreadSequence;
-        var lineEndPosition = currentSlice.PositionOfAny(ByteCR, ByteLF);
+        var array = new byte[span.Length + (lineFeedOnly ? 1 : 2)];
 
-        if (lineEndPosition == null)
-        {
-            // Not there.
-            return -1;
-        }
-
-        SequencePosition lineEnd;
-        scoped ReadOnlySpan<byte> headerSpan;
-        if (currentSlice.Slice(reader.Position, lineEndPosition.Value).Length == currentSlice.Length - 1)
-        {
-            // No enough data, so CRLF can't currently be there.
-            // However, we need to check the found char is CR and not LF
-
-            // Advance 1 to include CR/LF in lineEnd
-            lineEnd = currentSlice.GetPosition(1, lineEndPosition.Value);
-            var header = currentSlice.Slice(reader.Position, lineEnd);
-            headerSpan = header.IsSingleSegment ? header.FirstSpan : header.ToArray();
-            if (headerSpan[^1] != ByteCR)
-            {
-                RejectRequestHeader(headerSpan);
-            }
-            return -1;
-        }
-
-        // Advance 2 to include CR{LF?} in lineEnd
-        lineEnd = currentSlice.GetPosition(2, lineEndPosition.Value);
-        headerSpan = currentSlice.Slice(reader.Position, lineEnd).ToSpan();
-
-        if (headerSpan.Length < 5)
-        {
-            // Less than min possible headerSpan is 5 bytes a:b\r\n
-            RejectRequestHeader(headerSpan);
-        }
-
-        if (headerSpan[^2] != ByteCR)
-        {
-            // Sequence needs to be CRLF not LF first.
-            RejectRequestHeader(headerSpan[..^1]);
-        }
+        span.ToArray().CopyTo(array, 0);
```

Comment:
```suggestion
        span.CopyTo(array);
```
Is the allocation for `ToArray` needed? It can be copied directly instead?

The `array` is cated to span implicitely.